File: go/spore-core/tools/catalogue.go
// Standard Tool Catalogue (#81): the curated set of tools an architect drops
// into a harness, plus ready-made presets.
//
// Types
//   - StandardTool — a tool implementation bundled with its RegistryToolSchema
//     so the two can never be separated (issue #81, Q2). The HarnessBuilder's
//     Tool() destructures it into a registry Register() call.
//   - StandardTools — a namespace of one constructor per catalogue tool, each
//     returning a StandardTool, plus three presets: ReadonlySet, CodingSet,
//     and FullSet.
//
// Catalogue tools (constructor → registered name)
// Tier 1 (sandbox / stateless):
//   - ReadFile     → read_file     (EXISTING #5 tool)
//   - WriteFile    → write_file     (EXISTING)
//   - EditFile     → edit_file      (NEW)
//   - ListDir      → list_dir       (EXISTING)
//   - GrepFiles    → grep_files     (EXISTING)
//   - Grep         → grep           (NEW, output modes)
//   - FindFiles    → find_files     (EXISTING)
//   - BashCommand  → bash_command   (EXISTING)
//   - SendMessage  → send_message   (NEW)
//   - WebFetch     → web_fetch      (NEW)
//   - WebSearch    → web_search     (NEW)
//
// Tier 2 (storage via *ToolContext):
//   - TodoWrite    → todo_write     (NEW, RunStore key "todo")
//   - TaskList     → task_list      (EXISTING #71)
//
// Tier 3 (escalate / clarify):
//   - EnterPlanMode    → enter_plan_mode    (NEW)
//   - ExitPlanMode     → exit_plan_mode     (NEW)
//   - AskUserQuestion  → ask_user_question  (NEW)
//   - Abort            → abort              (NEW)
//
// Q5 — overlap with the EXISTING #5 catalogue (NO renames)
// The catalogue ships NET-NEW tools ALONGSIDE the existing #5 tools, never
// renaming them. Where a preset needs functionality an existing tool already
// provides, the preset REUSES the existing tool by its existing name:
//   - read_file, write_file, list_dir, find_files, grep_files, bash_command,
//     task_list are the EXISTING tools (their fixtures —
//     fixtures/tools/param_validation.json — stay byte-identical).
//   - edit_file and grep are NEW and live ALONGSIDE write_file / grep_files;
//     they do NOT replace them.
//
// Because StandardToolRegistry.Register is now a last-wins upsert (issue #81,
// Q1), registering a preset and then a custom tool of the same name lets the
// architect override a standard tool.
//
// Q3 — MemoryTool is DEFERRED. It depends on StorageScope (#79) and lands in a
// follow-on issue; it is intentionally NOT part of this catalogue.

package tools

import (
	sporecore "github.com/squirrelsoft-dev/spore-core/go/spore-core"
)

// StandardTool is the catalogue bundle (Tool implementation + RegistryToolSchema).
// It is defined in the root sporecore package (sporecore.StandardTool) so the
// HarnessBuilder can accept it without an import cycle; aliased here for ergonomic
// use from the tools package.
type StandardTool = sporecore.StandardTool

// schemaer is the optional Schema() method every catalogue tool implements.
type schemaer interface {
	Schema() sporecore.RegistryToolSchema
}

// newStandardTool bundles a tool that exposes a Schema() method.
func newStandardTool[T interface {
	sporecore.Tool
	schemaer
}](impl T) StandardTool {
	return StandardTool{Implementation: impl, Schema: impl.Schema()}
}

// StandardTools is the namespace of catalogue-tool constructors and presets.
// Unit type used as a namespace only — never instantiated.
type StandardTools struct{}

// ---- Tier 1 ---------------------------------------------------------------

// ReadFile — EXISTING #5 tool (Q5 overlap: reused, not renamed).
func (StandardTools) ReadFile() StandardTool { return newStandardTool(NewReadFileTool()) }

// WriteFile — EXISTING #5 tool (Q5 overlap: reused, not renamed).
func (StandardTools) WriteFile() StandardTool { return newStandardTool(NewWriteFileTool()) }

// EditFile — NEW unique-match in-place edit (alongside write_file).
func (StandardTools) EditFile() StandardTool { return newStandardTool(NewEditFileTool()) }

// ListDir — EXISTING #5 tool (Q5 overlap: reused, not renamed).
func (StandardTools) ListDir() StandardTool { return newStandardTool(NewListDirTool()) }

// GrepFiles — EXISTING #5 tool (Q5 overlap: reused, not renamed).
func (StandardTools) GrepFiles() StandardTool { return newStandardTool(NewGrepFilesTool()) }

// Grep — NEW regex search with output modes (alongside grep_files).
func (StandardTools) Grep() StandardTool { return newStandardTool(NewGrepTool()) }

// FindFiles — EXISTING #5 tool (Q5 overlap: reused, not renamed).
func (StandardTools) FindFiles() StandardTool { return newStandardTool(NewFindFilesTool()) }

// BashCommand — EXISTING #5 tool (Q5 overlap: reused, not renamed).
func (StandardTools) BashCommand() StandardTool { return newStandardTool(NewBashCommandTool()) }

// SendMessage — NEW; surfaces a UserMessage stream event via the harness loop.
func (StandardTools) SendMessage() StandardTool { return newStandardTool(NewSendMessageTool()) }

// WebFetch — NEW; GET a URL.
func (StandardTools) WebFetch() StandardTool { return newStandardTool(NewWebFetchTool()) }

// WebSearch — NEW; structured search over a configurable HTTP backend. The
// default has NO backend (calls error until one is configured); construct a
// StandardTool over NewWebSearchToolWithEndpoint to wire a real backend.
func (StandardTools) WebSearch() StandardTool { return newStandardTool(NewWebSearchTool()) }

// ---- Tier 2 ---------------------------------------------------------------

// TodoWrite — NEW; persists the todo list via RunStore key "todo".
func (StandardTools) TodoWrite() StandardTool { return newStandardTool(NewTodoWriteTool()) }

// TaskList — EXISTING #71 tool (Q5 overlap: reused, not renamed).
func (StandardTools) TaskList() StandardTool { return newStandardTool(NewTaskListTool()) }

// ---- Tier 3 ---------------------------------------------------------------

// EnterPlanMode — NEW; escalates HarnessSignal.EnterPlanMode.
func (StandardTools) EnterPlanMode() StandardTool { return newStandardTool(NewEnterPlanModeTool()) }

// ExitPlanMode — NEW; escalates HarnessSignal.ExitPlanMode{ Plan }.
func (StandardTools) ExitPlanMode() StandardTool { return newStandardTool(NewExitPlanModeTool()) }

// AskUserQuestion — NEW; returns ToolOutput.AwaitingClarification.
func (StandardTools) AskUserQuestion() StandardTool {
	return newStandardTool(NewAskUserQuestionTool())
}

// Abort — NEW; escalates HarnessSignal.Abort{ Reason }.
func (StandardTools) Abort() StandardTool { return newStandardTool(NewAbortTool()) }

// ---- Presets --------------------------------------------------------------

// ReadonlySet is the read-only investigation set: no mutating or escalating
// tools. Reuses the EXISTING read-only #5 tools by name (Q5 overlap) plus the
// NEW grep.
func (s StandardTools) ReadonlySet() []StandardTool {
	return []StandardTool{
		s.ReadFile(),
		s.ListDir(),
		s.GrepFiles(),
		s.Grep(),
		s.FindFiles(),
		s.WebFetch(),
		s.WebSearch(),
	}
}

// CodingSet is everything in ReadonlySet plus the mutating filesystem tools,
// shell, messaging, and the storage-backed todo/task tools. Reuses EXISTING
// tool names on overlap (Q5).
func (s StandardTools) CodingSet() []StandardTool {
	return []StandardTool{
		s.ReadFile(),
		s.WriteFile(),
		s.EditFile(),
		s.ListDir(),
		s.GrepFiles(),
		s.Grep(),
		s.FindFiles(),
		s.BashCommand(),
		s.SendMessage(),
		s.WebFetch(),
		s.WebSearch(),
		s.TodoWrite(),
		s.TaskList(),
	}
}

// FullSet is the CodingSet plus every Tier-3 control tool (plan / clarify /
// abort).
func (s StandardTools) FullSet() []StandardTool {
	set := s.CodingSet()
	set = append(set,
		s.EnterPlanMode(),
		s.ExitPlanMode(),
		s.AskUserQuestion(),
		s.Abort(),
	)
	return set
}

File: python/packages/spore_tools/src/spore_tools/tools/catalogue.py
"""Standard Tool Catalogue (#81): the curated set of tools an architect drops
into a harness, plus ready-made presets.

Mirrors ``rust/crates/spore-core/src/tools/catalogue.rs``.

Types
-----
* :class:`StandardTool` — a tool implementation bundled with its
  :class:`~spore_core.tool_registry.ToolSchema` so the two can never be
  separated (issue #81, Q2). ``HarnessBuilder.tool()`` destructures it.
* :class:`StandardTools` — a namespace of one constructor per catalogue tool,
  each returning a :class:`StandardTool`, plus three presets:
  :meth:`StandardTools.readonly_set`, :meth:`StandardTools.coding_set`, and
  :meth:`StandardTools.full_set`.

Catalogue tools (constructor → registered name)
-----------------------------------------------
Tier 1 (sandbox / stateless):

* :meth:`read_file` → ``read_file`` (EXISTING #5 tool)
* :meth:`write_file` → ``write_file`` (EXISTING)
* :meth:`edit_file` → ``edit_file`` (NEW)
* :meth:`list_dir` → ``list_dir`` (EXISTING)
* :meth:`grep_files` → ``grep_files`` (EXISTING)
* :meth:`grep` → ``grep`` (NEW, output modes)
* :meth:`find_files` → ``find_files`` (EXISTING)
* :meth:`bash_command` → ``bash_command`` (EXISTING)
* :meth:`send_message` → ``send_message`` (NEW)
* :meth:`web_fetch` → ``web_fetch`` (NEW)
* :meth:`web_search` → ``web_search`` (NEW)

Tier 2 (storage via ``ToolContext``):

* :meth:`todo_write` → ``todo_write`` (NEW, RunStore key ``"todo"``)
* :meth:`task_list` → ``task_list`` (EXISTING #71)
* :meth:`memory` → ``memory`` (NEW #82, scope-aware MemoryStore seam #78)

Tier 3 (escalate / clarify):

* :meth:`enter_plan_mode` → ``enter_plan_mode`` (NEW)
* :meth:`exit_plan_mode` → ``exit_plan_mode`` (NEW)
* :meth:`ask_user_question` → ``ask_user_question`` (NEW)
* :meth:`abort` → ``abort`` (NEW)

Q5 — overlap with the EXISTING #5 catalogue (NO renames)
--------------------------------------------------------
The catalogue deliberately ships NET-NEW tools ALONGSIDE the existing #5 tools,
never renaming them. Where a preset needs functionality that an existing tool
already provides, the preset REUSES the existing tool by its existing name:

* ``read_file``, ``write_file``, ``list_dir``, ``find_files``, ``grep_files``,
  ``bash_command`` are the EXISTING tools (their fixtures —
  ``fixtures/tools/param_validation.json`` — stay byte-identical).
* ``edit_file`` and ``grep`` are NEW and live ALONGSIDE ``write_file`` /
  ``grep_files``; they do NOT replace them.

Because :class:`~spore_core.tool_registry.StandardToolRegistry.register` is now
a last-wins upsert (issue #81, Q1), registering a preset and then a custom tool
of the same name lets the architect override a standard tool.

Q3 — MemoryTool (now LANDED, #82)
---------------------------------
``MemoryTool`` (``memory``) was deferred from #81 (it depended on the scoped
``MemoryStore`` seam from #78). It now ships here as a Tier-2 storage tool,
included in :meth:`coding_set` / :meth:`full_set` alongside ``task_list`` /
``todo_write``.
"""

from __future__ import annotations

from dataclasses import dataclass

from spore_core.tool_registry import Tool, ToolSchema

from .control import (
    AbortTool,
    AskUserQuestionTool,
    EnterPlanModeTool,
    ExitPlanModeTool,
)
from .edit import EditFileTool
from .exec import BashCommandTool
from .fs import ListDirTool, ReadFileTool, WriteFileTool
from .memory import MemoryTool
from .message import SendMessageTool
from .search import FindFilesTool, GrepFilesTool, GrepTool
from .tasklist import TaskListTool
from .todo import TodoWriteTool
from .web import WebFetchTool, WebSearchTool


@dataclass
class StandardTool:
    """A catalogue tool: its :class:`~spore_core.tool_registry.Tool`
    implementation bundled with its
    :class:`~spore_core.tool_registry.ToolSchema` so they can never drift apart
    (issue #81, Q2). ``HarnessBuilder.tool()`` destructures it."""

    implementation: Tool
    schema: ToolSchema


class StandardTools:
    """Namespace of catalogue-tool constructors and presets. Never
    instantiated — methods are ``@staticmethod`` (mirrors the Rust unit-struct
    namespace)."""

    # ---- Tier 1 ---------------------------------------------------------

    @staticmethod
    def read_file() -> StandardTool:
        """``read_file`` — EXISTING #5 tool (Q5 overlap: reused, not renamed)."""
        return StandardTool(ReadFileTool(), ReadFileTool.schema())

    @staticmethod
    def write_file() -> StandardTool:
        """``write_file`` — EXISTING #5 tool (Q5 overlap: reused, not renamed)."""
        return StandardTool(WriteFileTool(), WriteFileTool.schema())

    @staticmethod
    def edit_file() -> StandardTool:
        """``edit_file`` — NEW unique-match in-place edit (alongside ``write_file``)."""
        return StandardTool(EditFileTool(), EditFileTool.schema())

    @staticmethod
    def list_dir() -> StandardTool:
        """``list_dir`` — EXISTING #5 tool (Q5 overlap: reused, not renamed)."""
        return StandardTool(ListDirTool(), ListDirTool.schema())

    @staticmethod
    def grep_files() -> StandardTool:
        """``grep_files`` — EXISTING #5 tool (Q5 overlap: reused, not renamed)."""
        return StandardTool(GrepFilesTool(), GrepFilesTool.schema())

    @staticmethod
    def grep() -> StandardTool:
        """``grep`` — NEW regex search with output modes (alongside ``grep_files``)."""
        return StandardTool(GrepTool(), GrepTool.schema())

    @staticmethod
    def find_files() -> StandardTool:
        """``find_files`` — EXISTING #5 tool (Q5 overlap: reused, not renamed)."""
        return StandardTool(FindFilesTool(), FindFilesTool.schema())

    @staticmethod
    def bash_command() -> StandardTool:
        """``bash_command`` — EXISTING #5 tool (Q5 overlap: reused, not renamed)."""
        return StandardTool(BashCommandTool(), BashCommandTool.schema())

    @staticmethod
    def send_message() -> StandardTool:
        """``send_message`` — NEW; surfaces a ``StreamUserMessage`` via the loop."""
        return StandardTool(SendMessageTool(), SendMessageTool.schema())

    @staticmethod
    def web_fetch() -> StandardTool:
        """``web_fetch`` — NEW; GET a URL."""
        return StandardTool(WebFetchTool(), WebFetchTool.schema())

    @staticmethod
    def web_search() -> StandardTool:
        """``web_search`` — NEW; structured search over a configurable HTTP
        backend. The default has NO backend (calls error until one is
        configured); construct a :class:`StandardTool` over
        :meth:`WebSearchTool.with_endpoint` to wire a real backend."""
        return StandardTool(WebSearchTool(), WebSearchTool.schema())

    @staticmethod
    def web_search_with_endpoint(endpoint: str) -> StandardTool:
        """``web_search`` wired to a concrete backend endpoint. The plain
        :meth:`web_search` preset ships with no backend and errors on every
        call until one is configured; use this when you have a search endpoint
        (e.g. a Brave/Tavily-compatible URL) to POST the query to."""
        return StandardTool(WebSearchTool.with_endpoint(endpoint), WebSearchTool.schema())

    # ---- Tier 2 ---------------------------------------------------------

    @staticmethod
    def todo_write() -> StandardTool:
        """``todo_write`` — NEW; persists the todo list via RunStore key ``"todo"``."""
        return StandardTool(TodoWriteTool(), TodoWriteTool.schema())

    @staticmethod
    def task_list() -> StandardTool:
        """``task_list`` — EXISTING #71 tool (Q5 overlap: reused, not renamed)."""
        return StandardTool(TaskListTool(), TaskListTool.schema())

    @staticmethod
    def memory() -> StandardTool:
        """``memory`` — NEW #82; scope-aware read/write over the
        :class:`~spore_core.storage.MemoryStore` seam (#78)."""
        return StandardTool(MemoryTool(), MemoryTool.schema())

    # ---- Tier 3 ---------------------------------------------------------

    @staticmethod
    def enter_plan_mode() -> StandardTool:
        """``enter_plan_mode`` — NEW; escalates ``HarnessSignalEnterPlanMode``."""
        return StandardTool(EnterPlanModeTool(), EnterPlanModeTool.schema())

    @staticmethod
    def exit_plan_mode() -> StandardTool:
        """``exit_plan_mode`` — NEW; escalates ``HarnessSignalExitPlanMode { plan }``."""
        return StandardTool(ExitPlanModeTool(), ExitPlanModeTool.schema())

    @staticmethod
    def ask_user_question() -> StandardTool:
        """``ask_user_question`` — NEW; returns ``ToolOutputAwaitingClarification``."""
        return StandardTool(AskUserQuestionTool(), AskUserQuestionTool.schema())

    @staticmethod
    def abort() -> StandardTool:
        """``abort`` — NEW; escalates ``HarnessSignalAbort { reason }``."""
        return StandardTool(AbortTool(), AbortTool.schema())

    # ---- Presets --------------------------------------------------------

    @staticmethod
    def readonly_set() -> list[StandardTool]:
        """Read-only investigation set: no mutating or escalating tools. Reuses
        the EXISTING read-only #5 tools by name (Q5 overlap) plus the NEW
        ``grep``."""
        return [
            StandardTools.read_file(),
            StandardTools.list_dir(),
            StandardTools.grep_files(),
            StandardTools.grep(),
            StandardTools.find_files(),
            StandardTools.web_fetch(),
            StandardTools.web_search(),
        ]

    @staticmethod
    def coding_set() -> list[StandardTool]:
        """Coding set: everything in :meth:`readonly_set` plus the mutating
        filesystem tools, shell, messaging, and the storage-backed todo/task
        tools. Reuses EXISTING tool names on overlap (Q5)."""
        return [
            StandardTools.read_file(),
            StandardTools.write_file(),
            StandardTools.edit_file(),
            StandardTools.list_dir(),
            StandardTools.grep_files(),
            StandardTools.grep(),
            StandardTools.find_files(),
            StandardTools.bash_command(),
            StandardTools.send_message(),
            StandardTools.web_fetch(),
            StandardTools.web_search(),
            StandardTools.todo_write(),
            StandardTools.task_list(),
            StandardTools.memory(),
        ]

    @staticmethod
    def full_set() -> list[StandardTool]:
        """Full set: the :meth:`coding_set` plus every Tier-3 control tool
        (plan / clarify / abort)."""
        return [
            *StandardTools.coding_set(),
            StandardTools.enter_plan_mode(),
            StandardTools.exit_plan_mode(),
            StandardTools.ask_user_question(),
            StandardTools.abort(),
        ]


__all__ = ["StandardTool", "StandardTools"]
